File: daily_python/1751_Maximum_Number_of_Events_That_Can_Be_Attended_II.py
import os
import sys
from typing import List
import bisect

class Solution:
    def maxValue(self, events: List[List[int]], k: int) -> int:
        n = len(events)
        events.sort(key = lambda x : x[0])
        start = [event[0] for event in events]
        next_event = [bisect.bisect_right(start, event[1]) for event in events]
        dp = [[0] * n for _ in range(k)]
        dp[0][n-1] = events[n-1][2]
        for j in range(n-2, -1, -1):
            dp[0][j] = max(events[j][2], dp[0][j+1])
        for i in range(1, k):
            dp[i][n-1] = dp[i-1][n-1]
            for j in range(n-2, -1, -1):
                take = events[j][2] + (dp[i-1][next_event[j]] if next_event[j] < n else 0)
                dp[i][j] = max(dp[i][j+1], take)
        return max([max(r) for r in dp])

def readDataSet(filename):
    dataset = []
    with open(filename, 'r') as file:
        content = file.read().strip()
        blocks = content.split('\n\n')
        for block in blocks:
            lines = block.split('\n')
            events = [list(map(int, event.split(','))) for event in lines[0].split('=')[1].strip()[2:-3].split('},{')]
            k = int(lines[1].split('=')[1].strip()[:-1])
            dataset.append((events, k))
    return dataset

if __name__ == '__main__':
    if len(sys.argv) == 1:
        filename = os.path.splitext(__file__)[0] + '.txt'
    else:
        filename = sys.argv[1]
    dataset = readDataSet(filename)
    results = []
    solution = Solution()
    for events, k in dataset:
        results.append(solution.maxValue(events, k))
    for index, result in enumerate(results):
        print(f'Example {index + 1} : {result}')

File: daily/2025/1751_Maximum_Number_of_Events_That_Can_Be_Attended_II.cpp
#include <iostream>
#include <vector>

using namespace std;

class Solution {
public:
    int maxValue(vector<vector<int>>& events, int k) {
        int n = events.size();
        auto cmp = [](vector<int>& a, vector<int>& b){
            return a[0] < b[0];
        };
        sort(events.begin(), events.end(), cmp);
        vector<int> start_time;
        for (auto& event : events) start_time.push_back(event[0]);
        vector<int> next_event(n);
        for (int i = 0; i < n; i++) next_event[i] = upper_bound(start_time.begin(), start_time.end(), events[i][1]) - start_time.begin();
        vector<vector<int>> dp(k, vector<int>(n, 0));
        dp[0][n-1] = events[n-1][2];
        int ans = dp[0][n-1];
        for (int j = n-2; j >= 0; j--){
            dp[0][j] = max(events[j][2], dp[0][j+1]);
            ans = max(ans, dp[0][j]);
        }
        for (int i = 1; i < k; i++){
            dp[i][n-1] = dp[i-1][n-1];
            for (int j = n-2; j >= 0; j--){
                int take = events[j][2] + ((next_event[j] < n)? dp[i-1][next_event[j]] : 0);
                dp[i][j] = max(take, dp[i][j+1]);
                ans = max(ans, dp[i][j]);
            }
        }
        return ans;
    }
};

int main(){
    vector<vector<int>> events = {{1,2,4},{3,4,3},{2,3,1}};
    int k = 2;
    // vector<vector<int>> events = {{1,2,4},{3,4,3},{2,3,10}};
    // int k = 2;
    // vector<vector<int>> events = {{1,1,1},{2,2,2},{3,3,3},{4,4,4}};
    // int k = 3;
    // vector<vector<int>> events = {{74,91,40}};
    // int k = 1;
    Solution S;
    cout << S.maxValue(events, k) << endl;
    return 0;
}